src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -774,11 +855,20 @@ StartDistributedExecution(DistributedExecution *execution)
 
 	if (MultiShardCommitProtocol != COMMIT_PROTOCOL_BARE)
 	{
-		if (DistributedExecutionRequiresRollback(execution))
+		/*
```

Comment:
Question: Do we really need this `|| LocalExecutionHappened `? I added this during the development  because in my first iterations, `StartDistributedExecution ()` was called after local execution. But if we call   `StartDistributedExecution ()`, it should take care  of setting 2PC etc.

Reply:
Oh, yes  we need this. I had already added an example which needs  this:

```SQL
-- the same transaction block, but this time
-- enable local execution as well. The first
-- command is locally executed, the second
-- is remote, so 1 entry is expected
BEGIN;
INSERT INTO test_recovery VALUES ('hello');
INSERT INTO test_recovery VALUES ('world');
COMMIT;
SELECT count(*) FROM pg_dist_transaction;
 count 
-------
     1
(1 row)

SELECT recover_prepared_transactions();
 recover_prepared_transactions 
-------------------------------
                             0
(1 row)
```